File: android/src/main/kotlin/com/emarsys/emarsys_sdk/mapper/GeofenceMapper.kt
package com.emarsys.emarsys_sdk.mapper

import com.emarsys.emarsys_sdk.util.JsonUtils
import com.emarsys.mobileengage.api.geofence.Geofence
import com.emarsys.mobileengage.api.geofence.Trigger
import org.json.JSONObject

class GeofenceMapper : Mapper<List<Geofence>, List<Map<String, Any>>> {

    override fun map(result: List<Geofence>): List<Map<String, Any>> {
        return result
            .map {
                val resultMap = mutableMapOf(
                    "id" to it.id,
                    "lat" to it.lat,
                    "lon" to it.lon,
                    "radius" to it.radius,
                    "triggers" to mapTriggers(it.triggers)
                )
                it.waitInterval?.let { waitInterval ->
                    resultMap["waitInterval"] = waitInterval
                }
                resultMap.toMap()
            }
    }

    private fun mapTriggers(triggers: List<Trigger>): List<Map<String, Any>> {
        return triggers
            .map {
                val mappedAction = mutableMapOf<String, Any>(
                    "id" to it.id,
                    "type" to it.type.name,
                    "loiteringDelay" to it.loiteringDelay,
                )
                mapAction(it.action)?.let { action ->
                    mappedAction["action"] = action
                }
                mappedAction
            }
            .toList()
    }

    private fun mapAction(action: JSONObject): Map<String, Any>? {
        if (action.length() != 0) {
            val result = mutableMapOf(
                "id" to action.get("id"),
                "type" to action.get("type")
            )

            when (action["type"]) {
                "OpenExternalUrl" -> result["url"] = action.get("url")
                "MECustomEvent", "MEAppEvent" -> {
                    result["name"] = action.get("name")
                    result["payload"] = JsonUtils.toMap(action.get("payload") as JSONObject)
                }
            }
            return result.toMap()
        }
        return null
    }
}

File: lib/mappers/geofence_mapper.dart
import 'package:emarsys_sdk/mappers/mapper.dart';
import 'package:emarsys_sdk/model/geofence_model.dart';
import 'package:emarsys_sdk/model/geofence_trigger.dart';

class GeofenceMapper
    extends Mapper<List<Map<String, dynamic>?>, List<GeofenceModel>> {
  @override
  List<GeofenceModel> map(List<Map<String, dynamic>?> input) {
    return input
        .where((element) => element != null && element.isNotEmpty)
        .map((geofenceMap) => GeofenceModel(
            id: geofenceMap!["id"] as String,
            lat: geofenceMap["lat"] as double,
            lon: geofenceMap["lon"] as double,
            radius: geofenceMap["radius"] as double,
            waitInterval: geofenceMap["waitInterval"] as double?,
            triggers: mapTriggers(geofenceMap["triggers"])))
        .toList();
  }

  List<GeofenceTrigger> mapTriggers(List<dynamic> input) {
    return input.map((triggerMap) {
      final action = triggerMap["action"];
      return GeofenceTrigger(
          id: triggerMap["id"] as String,
          type: triggerMap["type"] as String,
          loiteringDelay: triggerMap["loiteringDelay"] as int,
          action: action == null ? null : Map<String, dynamic>.from(action));
    }).toList();
  }
}
